### PR TITLE
CIP-64 / CIP-66 compatible `TransactionArgs`

### LINKED ISSUE #109: Ensure TransactionArgs works with cip-64 and cip-66
# Rationale

In celo-blockchain transactions can be sent as arguments, and [internal/ethapi/transaction_args.go](https://github.com/celo-org/op-geth/blob/celo4/internal/ethapi/transaction_args.go#L345) creates the proper transaction type, and signed from an unlocked account. This is currently missing in op-geth.

# Implementation

Implementation should be analog to celo-blockchain's implementation.

### PR DESCRIPTION
**Pull request by ezdac**
Fixes #109

#### Add CIP-66 transaction type

This introduces the `CeloDenominatedTx` type, which will [when implemented](https://github.com/celo-org/op-geth/issues/64) allow for the fee-currency to be denominated in cel2 native token.

#### Make the `TransactionArgs` CIP-64/CIP-66 compatible
The `TransactionArgs` are used in some API endpoints to fill incomplete fields of passed in transactions and then 
convert this internally to an EVM-message or `Transaction`. This PR adds code that distinguishes some combination of passed in fields for the `TransactionArgs` between CIP-64 and CIP-66 transactions in order to create the concrete internal transaction type. The filling of missing required fields now considers wether the transaction is Celo-denominated or not.
Additionally, the new `MaxFeeInFeeCurrency` field is passed along to the internal transaction representations.

### CODE REVIEW

File: core/types/transaction.go [hunk 1]
```diff
@@ -110,6 +111,7 @@ type TxData interface {
 
 	// Celo specific fields
 	feeCurrency() *common.Address
+	maxFeeInFeeCurrency() *big.Int
```
Comment (karlb) [inline core/types/transaction.go:114]: IIRC @piersy suggested that we can avoid extending TxData by checking the tx type and casting to a specific tx type whenever we want to use a tx-type-specific attribute. I think that can work, but for consistency it should either be done for `feeCurrency` too or not at all.
I think going with the current approach is fine for now. I just wanted to bring up the point to see if there are any strong opionions or other thoughs about it.

File: core/types/transaction.go [hunk 2]
```diff
@@ -628,6 +632,11 @@ func (tx *Transaction) FeeCurrency() *common.Address {
 	return copyAddressPtr(tx.inner.feeCurrency())
 }
 
+// MaxFeeInFeeCurrency returns the max fee in the fee_currency for celo denominated txs.
```
Comment (karlb) [inline core/types/transaction.go]: The comment feels too much like the variable name. We can add some more information here.

Suggestion:
> MaxFeeInFeeCurrency is only used to guard against very quickly changing exchange rates. Txs must be discarded if MaxFeeInFeeCurrency is exceeded.

File: core/state_transition.go [hunk 3]
```diff
@@ -154,14 +154,15 @@ func toWordSize(size uint64) uint64 {
...
-	GasTipCap     *big.Int
+	To        *common.Address
+	From      common.Address
+	Nonce     uint64
+	Value     *big.Int
+	GasLimit  uint64
+	GasPrice  *big.Int
+	GasFeeCap *big.Int
+	GasTipCap *big.Int
+
```
Comment (karlb) [inline core/state_transition.go]: This looks like an unnecessary change.

File: core/state_transition.go [hunk 4]
```diff
@@ -213,7 +213,7 @@ func TransactionToMessage(tx *types.Transaction, s types.Signer, baseFee *big.In
 	}
 	// If baseFee provided, set gasPrice to effectiveGasPrice.
 	if baseFee != nil {
-		if msg.FeeCurrency != nil {
+		if tx.Type() == types.CeloDynamicFeeTxType {
```
Comment (karlb) [inline core/state_transition.go:215]: Is this change safe for the case where `tx.Type() == types.CeloDynamicFeeTxType` but `msg.FeeCurrency == nil`? Unfortunately, we didn't forbid `nil` currencies for CIP-64.

File: internal/ethapi/transaction_args.go [hunk 5]
```diff
@@ -445,10 +445,11 @@ func (args *TransactionArgs) ToMessage(globalGasCap uint64, baseFee *big.Int, ex
...
 	var (
-		gasPrice   *big.Int
-		gasFeeCap  *big.Int
-		gasTipCap  *big.Int
-		blobFeeCap *big.Int
+		gasPrice            *big.Int
+		gasFeeCap           *big.Int
+		gasTipCap           *big.Int
+		blobFeeCap          *big.Int
+		maxFeeInFeeCurrency *big.Int
```
Comment (karlb) [inline internal/ethapi/transaction_args.go:445]: Adding a `// Celo specific` comment in between will reduce the diff size because it breaks the formatter's alignment into two parts. Not sure if that is overall better in this case.

**Review by karlb — COMMENTED** [review]
Looks fine to me. Just some minor comments.

Will this allow creating CIP-64 txs from the JS console, or does that require additional changes?

File: core/state_transition.go (hunk 4 above)
Comment (ezdac) [reply core/state_transition.go:215]: Yes, the `ConvertGoldToCurrency` function just returns the passed in value in that case (here `baseFee`, which has the same effect as omitting the if case).

File: core/types/transaction.go (hunk 1 above)
Comment (ezdac) [reply core/types/transaction.go:114]: I generally don't like the concept of the TxData interface exposing the union of all struct fields.
If I would contribute to upstream, I would vote to keep the changes of this PR as is for the purpose of consistency though. 
When considering minimizing the diff and stronger segregation of Celo specific features for our fork I tend to lean towards the suggestion of @piersy.


So do you suggest implementing this in this PR?

File: internal/ethapi/transaction_args.go (hunk 5 above)
Comment (ezdac) [reply internal/ethapi/transaction_args.go:445]: Done in dbb2c5759723

File: core/state_transition.go (hunk 3 above)
Comment (ezdac) [reply core/state_transition.go]: Done in dbb2c5759723

File: core/types/transaction.go (hunk 2 above)
Comment (ezdac) [reply core/types/transaction.go]: Done in dbb2c5759723

File: core/types/transaction.go (hunk 1 above)
Comment (karlb) [reply core/types/transaction.go:114]: No, let's keep it as it is. If we want to change it, we should do so in a separate PR.

**Review by karlb — APPROVED** [review]

**Comment by ezdac** [comment]
> Will this allow creating CIP-64 txs from the JS console, or does that require additional changes?

I'm not sure what JS console you mean, but in general, if you add an additional argument `feeCurrency` to calls like `eth_fillTransaction` and `eth_signTransaction` (local node with keystore) you should now get back a CIP-64 transaction with filled in defaults (see the added test).

However it seems that most client libraries construct their transactions by potentially making individual calls to `eth_gasPrice` etc, then rlp-encoding and signing locally and using `eth_sendRawTransaction`:
https://github.com/celo-org/op-geth/issues/109#issuecomment-2084931456

**Comment by karlb** [comment]
> I'm not sure what JS console you mean

I meant to one you get with `geth --dev console`

**Comment by aaronmgdr** [comment]
@karlb can we talk about the `eth_fillTransaction` is it true that if i send params like 
`{feeCurrency: "0x....", value: 0, data: '0x...'}` it will give back  
`{feeCurrency: "0x....", value: 0, data: '0x...', maxFeePerGas: '0x', maxPriorityFeePerGas: '0x', maxFeeInFeeCurrency: '0x', gas: '0x' }`
